### PR TITLE
Fix `E231` bug: Inconsistent catch compared to pycodestyle, such as when dict nested in list

### LINKED ISSUE #10113: [E231] Inconsistent catch compared to pycodestyle, such as when dict nested in list
Searching through closed and opened issues, I did not see anywhere that this is intentional or known, apologies if I missed something. It appears that in certain scenarios E231 is not triggering as expected, whereas in pycodestyle is it being caught. It may impact other scenarios, but the primary one I could find is it not working on dicts inside a list. Strangely enough it does work in tuples, and outside lists, just not in lists. Even stranger, only if not the first key/value; it does work if it is the first key/value in the dict in the list.

**pyproject.toml**
```
[tool.ruff.lint]
select = [
  "E",
  "W",
]
```

**Python file**
```
"""Minimal repo."""


def main() -> None:
    """Primary function."""
    results = {
        "k1": [1],
        "k2":[2],
    }
    results_in_tuple = (
        {
            "k1": [1],
            "k2":[2],
        },
    )
    results_in_list = [
        {
            "k1": [1],
            "k2":[2],
        }
    ]
    results_in_list_first = [
        {
            "k2":[2],
        }
    ]


main()
```

**pycodestyle 2.11.1** catching all 4
```
$ pycodestyle --count min_repo.py
min_repo.py:8:13: E231 missing whitespace after ':'
min_repo.py:13:17: E231 missing whitespace after ':'
min_repo.py:19:17: E231 missing whitespace after ':'
min_repo.py:24:17: E231 missing whitespace after ':'
4
```

**ruff 0.2.2** catching 3/4
```
$ ruff --version
ruff 0.2.2
$ ruff check --preview min_repo.py
min_repo.py:8:13: E231 [*] Missing whitespace after ':'
   |
 6 |     results = {
 7 |         "k1": [1],
 8 |         "k2":[2],
   |             ^ E231
 9 |     }
10 |     results_in_tuple = (
   |
   = help: Add missing whitespace

min_repo.py:13:17: E231 [*] Missing whitespace after ':'
   |
11 |         {
12 |             "k1": [1],
13 |             "k2":[2],
   |                 ^ E231
14 |         },
15 |     )
   |
   = help: Add missing whitespace

min_repo.py:24:17: E231 [*] Missing whitespace after ':'
   |
22 |     results_in_list_first = [
23 |         {
24 |             "k2":[2],
   |                 ^ E231
25 |         }
26 |     ]
   |
   = help: Add missing whitespace

Found 3 errors.
[*] 3 fixable with the `--fix` option.
```

To note, if using ruff format, it would fix this, so not a blocker if you are also using format along with check. If you are only using ruff to replace pycodestyle without format (for whatever reason, such as your project not being ready/compatible with auto format), then you would hit this miss if only using ruff and not pycodestyle.
```
$ ruff format --check min_repo.py --diff
--- min_repo.py
+++ min_repo.py
@@ -5,23 +5,23 @@
     """Primary function."""
     results = {
         "k1": [1],
-        "k2":[2],
+        "k2": [2],
     }
     results_in_tuple = (
         {
             "k1": [1],
-            "k2":[2],
+            "k2": [2],
         },
     )
     results_in_list = [
         {
             "k1": [1],
-            "k2":[2],
+            "k2": [2],
         }
     ]
     results_in_list_first = [
         {
-            "k2":[2],
+            "k2": [2],
         }
     ]


1 file would be reformatted
```

I know E231 is still preview, and this is a pretty minor miss, but I assume the goal is for parity with pycodestyle (and consistency across nested objects).

### PR DESCRIPTION
**Pull request by augustelalande**
## Summary

Fix `E231` bug: Inconsistent catch compared to pycodestyle, such as when dict nested in list. Resolves #10113.

## Test Plan

Example from #10113 added to test fixture.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Thanks for looking into this. Using a stack here makes sense. I'm surprised that this isn't hitting performance more, considering that we're now allocating a new `Vec` for each logical line (actually two).

**Comment by augustelalande** [comment]
> Thanks for looking into this. Using a stack here makes sense. I'm surprised that this isn't hitting performance more, considering that we're now allocating a new `Vec` for each logical line (actually two).

Ya it's surprising. But is allocating a new `Vec` really significantly more expensive than allocating any other variable?

**Comment by augustelalande** [comment]
Took inspiration from `extraneous_whitespace.rs` and reduced it to one stack

**Comment by MichaReiser** [comment]
> Ya it's surprising. But is allocating a new Vec really significantly more expensive than allocating any other variable, if it's not populated?

That's correct: Creating an empty `Vec` is "free" because it doesn't allocate. But a `Vec` is different from e.g. a `TextSize` in that it allocates memory on the heap when you push the first element. That's different from a `TextSize`  variable that is always stored on the stack (which is as cheap as memory can be)

File: crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/missing_whitespace.rs [hunk 1]
```diff
@@ -97,7 +97,7 @@ pub(crate) fn missing_whitespace(line: &LogicalLine, context: &mut LogicalLinesC
                     if let Some(next_token) = iter.peek() {
                         match (kind, next_token.kind()) {
                             (TokenKind::Colon, _)
-                                if open_parentheses > 0 && prev_lsqb > prev_lbrace =>
+                                if matches!(brackets.last(), Some(TokenKind::Lsqb)) =>
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/missing_whitespace.rs:100]: I find that much easier to understand. Nice refactor

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**